File: Python/Algorithm/Is_Hamiltonian_Path.py
# Complexity: O(n^2 * 2^n), n = Number of Vertices

def hp(vr, grp):
    n = 1 << vr; dp = [[False] * n for _ in range(vr)]
    for i in range(vr): dp[i][1 << i] = True
    for i in range(n):
        for j in range(vr):
            if i & (1 << j):
                for k in range(vr):
                    if i & (1 << k) and (grp[k] == j or grp[j] == k) and j != k and dp[k][i ^ (1 << j)]:
                        dp[j][i] = True
                        break
    for i in range(vr):
        if dp[i][n - 1]: return True
    return False

def check(vr, edges):
    grp = [0] * vr
    for u, v in edges:
        u, v = u - 1, v - 1
        grp[u] = v; grp[v] = u
    return hp(vr, grp)

# Example usage
vr = 4  # Number of vertices
edges = [(1, 2), (2, 3), (3, 4)]  # Example edges
print(check(vr, edges))

File: CPP/Algorithm/Is_Hamiltonian_Path.cpp
// Complexity: O(n^2 * 2^n), n = Number of Vertices

int64_t gp[N], vr;
bool hp() {
	int n = (1ll << vr); bool dp[vr][n]{};
	for (int i = 0; i < vr; ++i) dp[i][(1ll << i)] = true;
	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < vr; ++j) {
			if (i & (1ll << j)) {
				for (int k = 0; k < vr; ++k) {
					if (i & (1ll << k) && (gp[k] == j || gp[j] == k) && j != k && dp[k][i^(1ll << j)]) {
						dp[j][i] = true; break;
					}
				}
			}
		}
	}
	for (int i = 0; i < vr; ++i) {
		if (dp[i][n - 1]) return true;
	}
	return false;
}

// Operation:
    int eg; cin >> vr >> eg;  // vr < 20
    for (i = 0; i < eg; ++i) {
        int u, v; cin >> u >> v;
        gp[--u] = --v; gp[v] = u;
    }
    cout << hp();